File: Proyecto Mis Finanzas Python/src/gestorAplicación/interno/usuario.py
from .suscripcion import Suscripcion
from .movimientos import Movimientos
from .corriente import Corriente
from .cuenta import Cuenta
from .ahorros import Ahorros
from .metas import Metas
from excepciones import banksException, usersException

class Usuario():
    #Atributos de clase
    _usuariosTotales = []

    #Constructor
    def __init__(self, **kwargs) -> None:
        #Atributos de instancia 
        Usuario._usuariosTotales.append(self)
        self.setId(len(Usuario.getUsuariosTotales()))
        self._bancosAsociados = []
        self._cuentasAhorroAsociadas = []
        self._cuentasCorrienteAsociadas = []
        self._cuentasAsociadas = []
        self._metasAsociadas = []
        self._contadorMovimientos = 0
        self._contadorMovimientosAux = 0
        self._movimientosAsociados = []
        self._limiteCuentas = 0
        for key in kwargs:
            if key == "_nombre":
                self.setNombre(kwargs[key])
            if key == "_correo":
                self.setCorreo(kwargs[key])
            if key == "_contrasena":
                self.setContrasena(kwargs[key])
            if key == "_suscripcion":
                self.setSuscripcion(kwargs[key])

    #Métodos de clase
    @classmethod
    def verificarCredenciales(cls, _nombre, _contrasena) -> object:
        for usuario in Usuario.getUsuariosTotales():
            if(usuario.getNombre() == _nombre or usuario.getCorreo() == _nombre):
                if(usuario.getContrasena() == _contrasena):
                    return usuario
        raise usersException.NoUserFoundException
    
    #Métodos de instancia
    def verificarContadorMovimientos(self):
        self.setContadorMovimientos(len(self.getMovimientosAsociados()))
        if(self.getContadorMovimientos() - self.getContadorMovimientosAux() == 5):
            self.setContadorMovimientosAux(self.getContadorMovimientosAux() + 5)
            match self.getSuscripcion():
                case Suscripcion.DIAMANTE:
                    self.setContadorMovimientos(0)
                    return("Felicidades, has alcanzado el nivel máximo de suscripción")
                case Suscripcion.ORO:
                    self.setContadorMovimientos(0)
                    self.setSuscripcion(Suscripcion.DIAMANTE)
                    return("Felicidades, has sido promovido al nivel de DIAMANTE, estos son tus beneficios: \nPuedes asociar un máximo de " + str(Suscripcion.DIAMANTE.getLimiteCuentas()) + " cuentas, la probabilidad de ganar en tu inversión es de " + str(Suscripcion.DIAMANTE.getProbabilidadInversion()))
                case Suscripcion.PLATA:
                    self.setContadorMovimientos(0)
                    self.setSuscripcion(Suscripcion.ORO)
                    return("Felicidades, has sido promovido al nivel de ORO, estos son tus beneficios: \nPuedes asociar un máximo de " + str(Suscripcion.ORO.getLimiteCuentas()) + " cuentas, la probabilidad de ganar en tu inversión es de " + str(Suscripcion.ORO.getProbabilidadInversion()))
                case Suscripcion.BRONCE:
                    self.setContadorMovimientos(0)
                    self.setSuscripcion(Suscripcion.PLATA)
                    return("Felicidades, has sido promovido al nivel de PLATA, estos son tus beneficios: \nPuedes asociar un máximo de " + str(Suscripcion.PLATA.getLimiteCuentas()) + " cuentas, la probabilidad de ganar en tu inversión es de " + str(Suscripcion.PLATA.getProbabilidadInversion()))
        else:
            return("Debes completar 5 movimientos para ser promovido de nivel, llevas " + str((self.getContadorMovimientos() - self.getContadorMovimientosAux())) + " movimiento(s)")
        
    def asociarBanco(self, banco) -> str:
        from gestorAplicación.externo.banco import Banco
        if(banco in Banco.getBancosTotales() and not(banco in self.getBancosAsociados())):
            self.getBancosAsociados().append(banco)
            return("El banco " + banco.getNombre() + " se ha asociado con éxito al usuario " + self.getNombre())
        else:
            return("No se encuentra el banco ó debes verificar que el banco que quieres asociar no se haya asociado antes, esta es la lista de bancos asociados: " + self.getBancosAsociados()[m] for m in range(0, len(self.getBancosAsociados())))
        
    def asociarCuenta(self, cuenta) -> str:
            if(not(cuenta in self.getCuentasAsociadas())):
                cuenta.setTitular(self)
                self.getCuentasAsociadas().append(cuenta)
                self.asociarBanco(cuenta.getBanco())
                if(isinstance(cuenta, Ahorros)):
                    return(self.asociarCuentaAhorros(cuenta))
                else:
                    if(cuenta.getCupo() == 0.0):
                        Corriente.inicializarCupo(cuenta)
                    return(self.asociarCuentaCorriente(cuenta))
            else:
                return("Debes comprobar que la cuenta no haya sido asociada con anterioridad.")
    
    def asociarMeta(self, meta) -> str:
        if(meta in Metas.getMetasTotales()):
            meta.setDueno(self)
            self.getMetasAsociadas().append(meta)
            return("La meta " + str(meta.getNombre()) + " se ha asociado con éxito al usuario " + self.getNombre())
        else:
            return("No se encuentra tu meta, debes verificar que la meta que quieres asociar exista" )
    
    def asociarMovimiento(self, movimiento) -> str:
        if(movimiento in Movimientos.getMovimientosTotales()):
            movimiento.setOwner(self)
            self.getMovimientosAsociados().append(movimiento)
            if(movimiento.getDestino() == None):
                return("El movimiento con origen " + movimiento.getOrigen().getNombre() + " ha sido asociada correctamente al usuario " + self.getNombre())
            else:
                if(not(movimiento.getDestino().getTitular() == self)):
                    movimiento.getDestino().getTitular().getMovimientosAsociados().append(movimiento)
                return("El movimiento con destino " + movimiento.getDestino().getNombre() + " ha sido asociada correctamente al usuario " + self.getNombre())
        else:
            return("No se encuentra el movimiento. Por favor asegurese de que el movimiento se haya realizado con éxito" )
    
    def asociarCuentaCorriente(self, corriente) -> str:
        if(corriente in Corriente.getCuentasCorrienteTotales()):
            self.getCuentasCorrienteAsociadas().append(corriente)
            return("La cuenta corriente " + corriente.getNombre() + " ha sido asociada correctamente al usuario " + self.getNombre())
        else:
            return("Debes verificar que la cuenta no haya sido asociada antes")
        
    def asociarCuentaAhorros(self, ahorros) -> str:
        if(ahorros in Ahorros.getCuentasAhorrosTotales()):
            self.getCuentasAhorroAsociadas().append(ahorros)
            return("La cuenta de ahorros " + ahorros.getNombre() + " ha sido asociada correctamente al usuario " + self.getNombre())
        else:
            return("Debes verificar que la cuenta no haya sido asociada antes")
    
    def mostrarBancosAsociados(self) -> object:
        bancos = self.getBancosAsociados()
        if(len(bancos) != 0):
            return bancos
        else:
            raise banksException.NoBanksException("Error. No hay bancos asociados a este usuario.")
        
    def mostrarCuentasAsociadas(self) -> object:
        cuentas = self.getCuentasAsociadas()
        if(len(cuentas) != 0):
            return cuentas
        else:
            return ("Primero debes asociar cuentas")
        
    def mostrarCuentasAhorroAsociadas(self) -> object:
        cuentas = self.getCuentasAhorroAsociadas()
        if(len(cuentas) != 0):
            return cuentas
        else:
            return ("Primero debes asociar cuentas")
        
    def mostrarCuentasCorrienteAsociadas(self) -> object:
        cuentas = self.getCuentasCorrienteAsociadas()
        if(len(cuentas) != 0):
            return cuentas
        else:
            return ("Primero debes asociar cuentas")
        
    # Métodos de la funcionalidad asesoramiento de inversiones
    def hallarUsuariogotaGota():
        contador = 0
        for i in range(len(Usuario.getUsuariosTotales())):
            if Usuario.getUsuariosTotales()[i].getNombre() == "gotaGota":
                contador = i
            else:
                continue
        return contador

    def hallarUsuarioImpuestosPortafolio():
        contador = 0
        for i in range(len(Usuario.getUsuariosTotales())):
            if Usuario.getUsuariosTotales()[i].getNombre() == "impuestosPortafolio":
                contador = i
            else:
                continue
        return contador

    def __str__(self) -> str:
        return("Usuario: " + self.getNombre() +
				"\nCorreo: " + self.getCorreo() +
				"\n#: " + str(self.getId()) +
				"\nCuentas Asociadas: " + str(self.getCuentasAsociadas()) +
				"\nSuscripción: " + str(self.getSuscripcion()))

    # Metodos funcionalidad de prestamos
# Metodos funcionalidad de prestamos
    def comprobarConfiabilidad(self):
        from .deuda import Deuda

        cuentasUsuario = self.getCuentasAhorroAsociadas()
        # conseguir la suscripcion
        suscripcion = self.getSuscripcion()
        # Comprobamos y contamos las deudas que estan asociadas al usuario
        deudasUsuario = Deuda.conseguirDeuda(self)
        if len(deudasUsuario)<suscripcion.getMaxDeudas():
            if len(cuentasUsuario) !=0:
                return cuentasUsuario
            else:
                return "¡Error! Usted no tiene ninguna cuenta Ahorros creada"
        else:
            return f"¡Error! La suscripción {suscripcion.name} solo permite realizar un total de {suscripcion.getMaxDeudas()}.Usted tiene {suscripcion.getMaxDeudas()}/{suscripcion.getMaxDeudas()}"
            ###########Falta imprimir las deudas

    #Métodos funcionalidad Compra de Cartera
    def retornarDeudas(self):
        cuentasConDeuda = []
        for cuenta in self.getCuentasAsociadas():
            if isinstance(cuenta, Corriente):
                if cuenta.getDisponible() != cuenta.getCupo():
                    cuentasConDeuda.append(cuenta)
        
        return cuentasConDeuda
    
    @staticmethod
    def capacidad_endeudamiento(cuentas, cuenta_a_aplicar):
        cuentas_capaces_deuda = []
        for cuenta in cuentas:
            if isinstance(cuenta, Corriente):
                deuda_validar = cuenta_a_aplicar.getCupo() - cuenta_a_aplicar.getDisponible()
                if cuenta.getDivisa() != cuenta_a_aplicar.getDivisa():
                    deuda_validar = Cuenta.dineroATenerDisponible(cuenta_a_aplicar, cuenta.getDivisa())
                if cuenta.getDisponible() >= deuda_validar and deuda_validar != 0:
                    cuentas_capaces_deuda.append(cuenta)
        
        return cuentas_capaces_deuda

    #Métodos Get & Set
    @classmethod
    def getUsuariosTotales(cls) -> list:
        return cls._usuariosTotales
    @classmethod
    def setUsuariosTotales(cls, _usuariosTotales) -> None:
        cls._usuariosTotales = _usuariosTotales
    def getNombre(self) -> str:
        return self._nombre
    def setNombre(self, _nombre) -> None:
        self._nombre = _nombre
    def getCorreo(self) -> str:
        return self._correo
    def setCorreo(self, _correo) -> None:
        self._correo = _correo
    def getContrasena(self) -> str:
        return self._contrasena
    def setContrasena(self, _contrasena) -> None:
        self._contrasena = _contrasena
    def getId(self) -> int:
        return self._id
    def setId(self, _id) -> None:
        self._id = _id
    def getSuscripcion(self) -> Suscripcion:
        return self._suscripcion
    def setSuscripcion(self, _suscripcion) -> None:
        self._suscripcion = _suscripcion
        self.setLimiteCuentas(self.getSuscripcion().getLimiteCuentas())
    def getCuentasAsociadas(self) -> list:
        return self._cuentasAsociadas
    def setCuentasAsociadas(self, _cuentasAsociadas) -> None:
        self._cuentasAsociadas = _cuentasAsociadas
    def getLimiteCuentas(self) -> int:
        return self._limiteCuentas
    def setLimiteCuentas(self, _limiteCuentas) -> None:
        self._limiteCuentas = _limiteCuentas
    def getBancosAsociados(self) -> list:
        return self._bancosAsociados
    def setBancosAsociados(self, _bancosAsociados) -> None:
        self._bancosAsociados = _bancosAsociados
    def getContadorMovimientos(self) -> int:
        return self._contadorMovimientos
    def setContadorMovimientos(self, _contadorMovimientos) -> None:
        self._contadorMovimientos = _contadorMovimientos
    def getMetasAsociadas(self) -> list:
        return self._metasAsociadas
    def setMetasAsociadas(self, _metasAsociadas) -> None:
        self._metasAsociadas = _metasAsociadas
    def getMovimientosAsociados(self) -> list:
        return self._movimientosAsociados
    def setMovimientosAsociados(self, _movimientosAsociados) -> None:
        self._movimientosAsociados = _movimientosAsociados
    def getCuentasCorrienteAsociadas(self) -> list:
        return self._cuentasCorrienteAsociadas
    def setCuentasCorrienteAsociadas(self, _cuentasCorrienteAsociadas) -> None:
        self._cuentasCorrienteAsociadas = _cuentasCorrienteAsociadas
    def getCuentasAhorroAsociadas(self) -> list:
        return self._cuentasAhorroAsociadas
    def setCuentasAhorroAsociadas(self, _cuentasAhorroAsociadas) -> None:
        self._cuentasAhorroAsociadas = _cuentasAhorroAsociadas
    def getContadorMovimientosAux(self) -> int:
        return self._contadorMovimientosAux
    def setContadorMovimientosAux(self, _contadorMovimientosAux) -> None:
        self._contadorMovimientosAux = _contadorMovimientosAux

File: Proyecto Mis Finanzas Java/src/gestorAplicación/interno/Usuario.java
package gestorAplicación.interno;

import java.io.Serializable;
import java.util.ArrayList;
import gestorAplicación.externo.Banco;
import gestorAplicación.externo.Estado;
import gestorAplicación.externo.Tablas;

public class Usuario implements Serializable {
	//Atributos
	//Atributos para serializar
	private static final long serialVersionUID = 3L;
	public static final String nombreD = "Usuarios";
	
	//Atributos de instancia
	private String nombre;
	private String correo;
	private String contrasena;
	private int id;
	private Suscripcion suscripcion;
	private static transient ArrayList<Usuario> usuariosTotales = new ArrayList<Usuario>();
	private ArrayList<Banco> bancosAsociados = new ArrayList<Banco>();
	
	//Funcionalidad de Suscripciones de Usuario
	private ArrayList<Cuenta> cuentasAsociadas = new ArrayList<Cuenta>();
	private int limiteCuentas;
	private int contadorMovimientos;
	private int contadorMovimientosAux;

	
	//Funcionalidad Asesor inversiones
	private ArrayList<Movimientos> movimientosAsociados = new ArrayList<Movimientos>();
	private ArrayList<Corriente> CuentasCorrienteAsociadas = new ArrayList<Corriente>();
	private ArrayList<Ahorros> CuentasAhorrosAsociadas = new ArrayList<Ahorros>();
	private ArrayList<Metas> metasAsociadas = new ArrayList<Metas>();

	//Constructores
	public Usuario(String nombre, String correo, String contrasena, Suscripcion suscripcion) {
		Usuario.getUsuariosTotales().add(this);
		this.setSuscripcion(suscripcion);
		this.setLimiteCuentas(suscripcion.getLimiteCuentas());
		this.setNombre(nombre);
		this.setContrasena(contrasena);
		this.setCorreo(correo);
		this.setId(Usuario.getUsuariosTotales().size());
	}
	
	public Usuario(String nombre, String correo, String contrasena) {
		Usuario.getUsuariosTotales().add(this);
		this.setSuscripcion(Suscripcion.BRONCE);
		this.setLimiteCuentas(this.getSuscripcion().getLimiteCuentas());
		this.setNombre(nombre);
		this.setContrasena(contrasena);
		this.setCorreo(correo);
		this.setId(Usuario.getUsuariosTotales().size());
	}
	//Constructor por defecto
	public Usuario() {
		this("Pepe Morales", "PepeMorales@mail.com", "12345", Suscripcion.DIAMANTE);
	}

	//Métodos de clase
	public static Object verificarCredenciales(String nombre, String contraseña) {
		for (Usuario usuario: usuariosTotales) {
			if (usuario.getNombre().equals(nombre) || usuario.getCorreo().equals(nombre)) {
				if (usuario.getContrasena().equals(contraseña)) {
					return usuario;
				}
			}
		}
		return null;
	}
	
	// Funcionalidad Asesor Inversiones
	public static int hallarUsuariogotaGota() {
		int contador = 0;
		for (int i = 0; i < Usuario.getUsuariosTotales().size(); i++) {
			if (Usuario.getUsuariosTotales().get(i).getNombre().equals("gotaGota")) {
				contador = i;
			} else {
				continue;
			}
		}
		return contador;
	}

	public static int hallarUsuarioImpuestosPortafolio() {
		int contador = 0;
		for (int i = 0; i < Usuario.getUsuariosTotales().size(); i++) {
			if (Usuario.getUsuariosTotales().get(i).getNombre().equals("impuestosPortafolio")) {
				contador = i;
			} else {
				continue;
			}
		}
		return contador;
	}
	
	//Métodos de instancia
	//Funcionalidad de Suscripciones de Usuarios
	public String verificarContadorMovimientos() {
		this.setContadorMovimientos(this.getMovimientosAsociados().size());
		if(this.getContadorMovimientos() - this.getContadorMovimientosAux() == 5) {
			this.setContadorMovimientosAux(this.getContadorMovimientosAux() + 5);
			switch(this.getSuscripcion()) {
				case DIAMANTE:
					this.setContadorMovimientos(0);
					return("Felicidades, has alcanzado el nivel máximo de suscripción");
				case ORO:
					this.setContadorMovimientos(0);
					this.setSuscripcion(Suscripcion.DIAMANTE);
					return("Felicidades, has sido promovido al nivel de DIAMANTE, estos son tus beneficios: " + "puedes asociar un máximo de " + Suscripcion.DIAMANTE.getLimiteCuentas() + " cuentas, la probabilidad de ganar en tu inversión es de " + Suscripcion.DIAMANTE.getProbabilidad_Inversion());
				case PLATA:
					this.setContadorMovimientos(0);
					this.setSuscripcion(Suscripcion.ORO);
					return("Felicidades, has sido promovido al nivel de ORO, estos son tus beneficios: " + "puedes asociar un máximo de " + Suscripcion.ORO.getLimiteCuentas() + " cuentas, la probabilidad de ganar en tu inversión es de " + Suscripcion.ORO.getProbabilidad_Inversion());
				case BRONCE:
					this.setContadorMovimientos(0);
					this.setSuscripcion(Suscripcion.PLATA);
					return("Felicidades, has sido promovido al nivel de PLATA, estos son tus beneficios: " + "puedes asociar un máximo de " + Suscripcion.PLATA.getLimiteCuentas() + " cuentas, la probabilidad de ganar en tu inversión es de " + Suscripcion.PLATA.getProbabilidad_Inversion());
				default:
					return("");
			
			}
		}else {
			return("Debes completar 5 movimientos para ser promovido de nivel, llevas " + (this.getContadorMovimientos()- this.getContadorMovimientosAux()) + " movimiento(s)");
		}
	}
	
	public String asociarBanco(Banco banco) {
		if(Banco.getBancosTotales().contains(banco) && !bancosAsociados.contains(banco)) {
			this.getBancosAsociados().add(banco);
			return("El banco " + banco.getNombre() + " se ha asociado con éxito al usuario " + this.getNombre());
		}else {
			return("No se encuentra el banco ó debes verificar que el banco que quieres asociar no se haya asociado antes, esta es la lista de bancos asociados: " + this.mostrarBancosAsociados());
		}
	}
	
	public String asociarCuenta(Cuenta cuenta) {
		if(!cuentasAsociadas.contains(cuenta)) {
			cuenta.setTitular(this);
			this.getCuentasAsociadas().add(cuenta);
			if(cuenta instanceof Ahorros) {
				return(this.asociarCuentaAhorros((Ahorros) cuenta));
			}else {
				if(((Corriente) cuenta).getCupo() == 0.0) {
					Corriente.inicializarCupo((Corriente) cuenta);
				}
				return(this.asociarCuentaCorriente((Corriente) cuenta));
			}
		}else {
			return("Debes comprobar que la cuenta no haya sido asociada con anterioridad.");
		}
	}
	
	public String asociarMeta(Metas meta) {
		if(Metas.getMetasTotales().contains(meta)) {
			meta.setDueno(this);
			this.getMetasAsociadas().add(meta);
			return("La meta " + meta.getNombre() + " se ha asociado con éxito al usuario " + this.getNombre());
		}else {
			return("No se encuentra tu meta, debes verificar que la meta que quieres asociar exista" );
		}
	}
	
	public String asociarMovimiento(Movimientos movimiento) {
		if(Movimientos.getMovimientosTotales().contains(movimiento)) {
			movimiento.setOwner(this);
			this.getMovimientosAsociados().add(movimiento);
			if (movimiento.getDestino() == null) {
				return("El movimiento con origen " + movimiento.getOrigen().getNombre() + " ha sido asociada correctamente al usuario " + this.getNombre());
			}else {
				if(!movimiento.getDestino().getTitular().equals(this)) {
					movimiento.getDestino().getTitular().getMovimientosAsociados().add(movimiento);
				}
				return("El movimiento con destino " + movimiento.getDestino().getNombre() + " ha sido asociada correctamente al usuario " + this.getNombre());
			}
		}else {
			return("No se encuentra el movimiento. Por favor asegurese de que el movimiento se haya realizado con éxito" );
		}
	}
	
	public String asociarCuentaCorriente(Corriente corriente) {
		if(Corriente.getCuentasCorrienteTotales().contains(corriente)) {
			this.getCuentasCorrienteAsociadas().add(corriente);
			return("La cuenta corriente " + corriente.getNombre() + " ha sido asociada correctamente al usuario " + this.getNombre());
		}else {
			return("Debes verificar que la cuenta no haya sido asociada antes");
		}
	}
	
	public String asociarCuentaAhorros(Ahorros ahorros) {
		if(Ahorros.getCuentasAhorroTotales().contains(ahorros)) {
			this.getCuentasAhorrosAsociadas().add(ahorros);
			return("La cuenta de ahorros " + ahorros.getNombre() + " ha sido asociada correctamente al usuario " + this.getNombre());
		}else {
			return("Debes verificar que la cuenta no haya sido asociada antes");
		}
	}
	
	public Object mostrarBancosAsociados() {
		ArrayList<Banco> bancos = this.getBancosAsociados();
		if(bancos.size() != 0) {
			return bancos;
		}else {
			return ("Primero debes asociar bancos");
		}
	}
	
	public Object mostrarCuentasAsociadas() {
		ArrayList<Cuenta> cuentas = this.getCuentasAsociadas();
		if(cuentas.size() != 0) {
			return cuentas;
		}else {
			return ("Primero debes asociar cuentas");
		}
	}
	
	//    Funcionalidad Prestamos
	public ArrayList<?> comprobarConfiabilidad(){
		//Deserializacion de las cuentas
		ArrayList<Ahorros> cuentasUsuario = this.getCuentasAhorrosAsociadas();
		ArrayList<String> cadena = new ArrayList<>();

//		Conseguimos la suscripciones y miramos las deudas
		Suscripcion suscripcion = getSuscripcion();
//		comprobamos y contamos las deudas que estan asociadas al usuario
		ArrayList<Deuda> deudasUsuario = (ArrayList<Deuda>) Deuda.conseguirDeudas(this);
//		returns
		if(deudasUsuario.size()<suscripcion.getMaxDeudas()){
			if(cuentasUsuario.size()!=0){
				return cuentasUsuario;
			}else{
				cadena.add("¡Error! Usted no tiene ninguna cuenta Ahorros creada");
			}
		}else{
			cadena.add("¡Error! La suscripción "+this.getSuscripcion().name()+" solo permite realizar un total de "+this.getSuscripcion().getMaxDeudas()+". Usted tiene"+this.getSuscripcion().getMaxDeudas()+"/"+this.getSuscripcion().getMaxDeudas()+" Deudas");
//			Agrega a cadena todas las cuentas del usuario para mostrarselas
			Tablas.impresionDeudas(deudasUsuario);
		}
		return cadena;
	}

	//Funcionalidad Compra Cartera
	public ArrayList<Corriente> retornarDeudas(){
		ArrayList<Corriente> cuentasConDeuda = new ArrayList<Corriente>();
		for (Cuenta cuenta: cuentasAsociadas) {
			if(cuenta instanceof Corriente) {
				if (((Corriente) cuenta).getDisponible().compareTo(((Corriente) cuenta).getCupo()) != 0) {
					cuentasConDeuda.add((Corriente) cuenta);
				}
			}
		}
		return cuentasConDeuda;
	}
	
	//Funcionalidad Compra Cartera
	public ArrayList<Corriente> Capacidad_Endeudamiento(ArrayList<Cuenta> cuentas, Corriente cuenta_a_Aplicar) {
//		double deuda = cuenta_a_Aplicar.getCupo() - cuenta_a_Aplicar.getDisponible();
		ArrayList<Corriente> cuentasCapacesDeuda = new ArrayList<Corriente>();
		for (Cuenta cuenta: cuentas) {
			if (cuenta instanceof Corriente) {
				double deudaValidar = ((Corriente) cuenta).getCupo() - ((Corriente) cuenta).getDisponible();
				if (!cuenta.getDivisa().equals(cuenta_a_Aplicar.getDivisa())) {
					deudaValidar = Cuenta.DineroaTenerDisponible(cuenta_a_Aplicar, cuenta.getDivisa());
				}
				if (((Corriente) cuenta).getDisponible() >= deudaValidar && deudaValidar != 0) {
					cuentasCapacesDeuda.add((Corriente) cuenta);
				}
//				if (((Corriente) cuenta).getDisponible() >= deuda) {
//					cuentasCapacesDeuda.add((Corriente) cuenta);
//				}
			}
		}
		return cuentasCapacesDeuda;
	}
	
	public void eliminarMetas(int n) {
		this.getMetasAsociadas().remove(n);
		Metas.getMetasTotales().remove(n);
	}
	
	public static void limpiarPropiedades(ArrayList<String> arreglo) {
		arreglo.remove("serialVersionUID");
		arreglo.remove("nombreD");
		arreglo.remove("usuariosTotales");
		arreglo.remove("bancosAsociados");
		arreglo.remove("cuentasAsociadas");
		arreglo.remove("movimientosAsociados");
		arreglo.remove("CuentasCorrienteAsociadas");
		arreglo.remove("CuentasAhorrosAsociadas");
		arreglo.remove("metasAsociadas");
		arreglo.remove("$SWITCH_TABLE$gestorAplicación$interno$Suscripcion");
	} 
	
	@Override
	protected void finalize() { System.out.println("El usuario con id: " + this.getId() + " y nombre: " + this.getNombre() + " fue eliminado satisfactoriamente del sistema."); }
	
	@Override
	public boolean equals(Object o) {
		if(this.getId() == ((Usuario) o).getId()){
			return true;
		}else {
			return false;
		}	
	}
	
	public String toString() {
		return "Usuario: " + this.getNombre() +
				"\nCorreo: " + this.getCorreo() +
				"\n#: " + this.getId() +
				"\nCuentas Asociadas: " + this.getCuentasAsociadas() +
				"\nSuscripción: " + this.getSuscripcion();
	}
	
	//Métodos Get & Set
	public static ArrayList<Usuario> getUsuariosTotales() { return usuariosTotales; }
	public static void  setUsuariosTotales(ArrayList<Usuario> usuariosTotales) { Usuario.usuariosTotales = usuariosTotales; }
	public String getNombre() { return nombre; }
	public void setNombre(String nombre) { this.nombre = nombre; }
	public String getCorreo() { return correo; }
	public void setCorreo(String correo) { this.correo = correo; }
	public String getContrasena() { return contrasena; }
	public void setContrasena(String contrasena) { this.contrasena = contrasena; }
	public int getId() { return id; }
	public void setId(int id) { this.id = id; }
	public Suscripcion getSuscripcion() { return suscripcion; }
	public void setSuscripcion(Suscripcion suscripcion) { this.suscripcion = suscripcion; }
	public ArrayList<Cuenta> getCuentasAsociadas() { return cuentasAsociadas; }
	public void setCuentasAsociadas(ArrayList<Cuenta> cuentasAsociadas) { this.cuentasAsociadas = cuentasAsociadas; }
	public int getLimiteCuentas() { return limiteCuentas; }
	public void setLimiteCuentas(int limiteCuentas) { this.limiteCuentas = limiteCuentas; }
	public ArrayList<Banco> getBancosAsociados() { return bancosAsociados; }
	public void setBancosAsociados(ArrayList<Banco> bancosAsociados) { this.bancosAsociados = bancosAsociados; }
	public int getContadorMovimientos() { return contadorMovimientos; }
	public void setContadorMovimientos(int contadorMovimientos) { this.contadorMovimientos = contadorMovimientos; }
	public ArrayList<Metas> getMetasAsociadas() {return metasAsociadas;}
	public void setMetasAsociadas(ArrayList<Metas> metasAsociadas) {this.metasAsociadas = metasAsociadas;}
	public ArrayList<Movimientos> getMovimientosAsociados() {return movimientosAsociados;}
	public void setMovimientosAsociados(ArrayList<Movimientos> movimientosAsociados) {this.movimientosAsociados = movimientosAsociados;}
	public ArrayList<Corriente> getCuentasCorrienteAsociadas() {return CuentasCorrienteAsociadas;}
	public void setCuentasCorrienteAsociadas(ArrayList<Corriente> cuentasCorrienteAsociadas) {CuentasCorrienteAsociadas = cuentasCorrienteAsociadas;}
	public ArrayList<Ahorros> getCuentasAhorrosAsociadas() {return CuentasAhorrosAsociadas;}
	public void setCuentasAhorrosAsociadas(ArrayList<Ahorros> cuentasAhorrosAsociadas) {CuentasAhorrosAsociadas = cuentasAhorrosAsociadas;}
	public int getContadorMovimientosAux() { return contadorMovimientosAux; }
	public void setContadorMovimientosAux(int contadorMovimientosAux) { this.contadorMovimientosAux = contadorMovimientosAux; }
}
